### ISSUE TITLE
Allow linking to a predefined set of selected starters

### ISSUE THREAD

**Issue by odrotbohm** [spring-io/start.spring.io#188]
In this [Twitter thread](https://twitter.com/rotnroll666/status/1148588327668523009) and some related discussion in #710 the idea of being able to link to a pre-defined set of starters shaped up.

The general use case is linking to those from our very own reference documentation and web content that explain users how to create a minimal sample project to start with to e.g. reproduce a bug or the like.

**Comment by gregturn** [comment]
I agree with having options on the website selectable via a deep link.

**Comment by sdeleuze** [comment]
That would also makes non-default-but-popular arrangement like Kotlin + Gradle more discoverable as previously discussed with @oodamien.

**Comment by philwebb** [comment]
This issue has generated quite a bit of discussion internally and I want to try and capture some of it here. Apologies if I've misunderstood or misrepresent what's been said.

First, it's perhaps worth capture some of the personas that are interested in this use-case. 

Three likely ones that we have are:

- An advocate for FooTech Inc that wants to blog about their Spring Boot integration
- An experienced Spring developer wanting to learn about FooTech
- An experienced FooTech developer wanting to learn about Spring

FooTech here is just a general placeholder for a technology, it could be Neo4J, Okta, MongoDB etc.

Each persona has a slightly different take on what they might to do:

- The FooTech advocate wants to get people setup with a project that works as quickly as possible. They don't really care too much about Spring, they just want to get FooTech in the users favorite IDE or editor as quickly as possible.

- The Spring developer already knows about start.spring.io. They most often use it directly from their IDE (not via the web). They know about the Spring Boot concepts such as starters and auto-configuration. They want to learn FooTech by following the blog written by the advocate.

- The FooTech developer wants to explore if Spring is a good vehicle for actually developing FooTech applications. They don't know anything about Spring Boot concepts and they have no idea what start.spring.io is. They don't know that their favorite IDE already has start.spring.io integration plugins installed.

I think these use-cases and personas are important because understanding all of them helps provide context about why https://github.com/spring-io/initializr/issues/710 was declined and perhaps why there's some resistance to just provided a "Download Project" friendly link. Summarizing some of the arguments:

- The Spring Developer can no longer use their IDE directly, they are forced into the web site.
- There's little incentive for the FooTech advocate to actually explain what's going on (although not having a link doesn't necessarily create the incentive)
- Links can go stale quickly. Users would end up downloading old versions of libraries.

Unfortunately, I think some of the behaviors that we're trying to drive by not providing links aren't necessarily happening. Specifically:

- The blog written by the FooTech advocate ends up just saying "go to the site and type web, devtools and FooTech into the search and click the green button". We don't end up with an explanation of what those things do and we don't tell the FooTech developer that we have nice IDE integration.
- Since the default version of Boot is the latest, it's possible the search terms may have changed and won't even work. The user gets no help or guidance.
- The FooTech advocate might get so frustrated that they just provide a curl link instead. We've effectively pushed them into the worst possible outcome for us.

I think most of us on the team agree that a direct link to the project zip doesn't help us create the experience we want. I think we need to investigate other options that actually help us achieve our goals. I'll add one proposal that myself and @snicoll discussed in the next comment. We should add more ideas as they come to us.

**Comment by philwebb** [comment]
After some brainstorming here's one proposal for how we could offer a better experience for all the personas above.

Starting with the FooTech advocate. They're writing their blog and they want to provide a nice getting started experience. To create the link they go to start.spring.io and add the dependencies that they need:

![page1](https://user-images.githubusercontent.com/519772/61068290-926c3b80-a401-11e9-8990-3a32b0abb34d.png)


Instead of clicking "Generate Project" they click "..." and pick "Share Project". This opens a small popup window a bit like the GitHub "Clone or Download" button with a URL they can copy.

The URL would be something like: `http://start.spring.io/project/2.1.x/java/maven?group=example&artifact=foo&dependencies=web,devtools,footech`

Some important aspects are encoded here:
- The link is relatively readable
- The link is not to `/project` not `/`
- Despite the user selecting `2.1.6` the link is for `2.1.x`
- Query parameters are used for the group/artifact, dependencies etc

The FooTech advocate can now include the link directly in their blog to be consumed by either the Spring or FooTech developer. Clicking the link gives something like this:

![page2](https://user-images.githubusercontent.com/519772/61068308-9dbf6700-a401-11e9-8fba-c1a6ee517d11.png)

With this page, we're in control again and we can provide information that we think is relevant. The onus is no longer on the FooTech advocate to explain things. Furthermore, were free to change this page even after a link is created.

Some ideas about what we could include:

- Information about the fact that the project was created from start.spring.io and is pre-configured
- A download button with the latest in the .x line. E.g. a link to 2.1.x would actually download 2.1.6
- Information about an available upgrade. We can offer a 1 click update to 2.2.x and deal with any missing dependencies in a nice way
- IDE import details. We can advertise all the nice features and possible even offer 1-click imports (like GitHub does if you have their app)
- We can sort of fix https://github.com/spring-io/initializr/issues/710 because here we can offer sensible curl links
- We can offer an edit button to allow users to change things if they want to

**Comment by wilkinsona** [comment]
Thanks for writing this up, Phil.

> IDE import details. We can advertise all the nice features and possible even offer 1-click imports (like GitHub does if you have their app)

This feels like something that we might want to offer on the "main" page too, particularly if we can get custom URL handler support or the like from the IDEs so that it's a one-click kind of thing. That said, we have less real estate available on the main page without things seeming cluttered. I do think it would feel a bit strange to only offer that capability for pre-configured projects and not for everyone though.

> We can offer an edit button to allow users to change things if they want to

This one gave me pause for thought. The current (and only) page that we have is designed to allow users to change things if they want, they're just starting from a blank canvas rather than something that's been preconfigured. Would the edit but take them to that page with things preselected? That feels like the right thing to do to me, but once you've provided that capability I find myself feeling that the lines between `start.spring.io/` and `start.spring.io/project` are becoming a bit blurred.

**Comment by philwebb** [comment]
> Would the edit but take them to that page with things preselected?

I've not really given it much thought. Perhaps we should iterate and start off without even having that as an option. The kind of use-case I imagined was wanting to change the build system, or group ID.

I was originally thinking about all that link content just being a box that appeared above the regular start.spring.io page, but I thought it would be easier in the end to have a different page altogether.

**Comment by snicoll** [comment]
> This feels like something that we might want to offer on the "main" page too, particularly if we can get custom URL handler support or the like from the IDEs so that it's a one-click kind of thing.

I agree, the same goes with the ability to "go back toe the same page" as @sdeleuze requested in the past to promote the ability for a user to change defaults (language, build system, etc). The same goes for promoting curl/httpie commands.

>  I find myself feeling that the lines between start.spring.io/ and start.spring.io/project are becoming a bit blurred.

I like the mock-up @philwebb designed and the fact that you end up on a page that states "this is what whoever created that link would like you to get started with". I also like the fact that, from there, you have to click an extra link to change stuff. I think this is a nice touch when being redirected from a blog post/tutorial.

**Comment by wilkinsona** [comment]
> I like the mock-up @philwebb designed and the fact that you end up on a page that states "this is what whoever created that link would like you to get started with". I also like the fact that, from there, you have to click an extra link to change stuff. I think this is a nice touch when being redirected from a blog post/tutorial.

I like that a lot too. Particularly as it creates a clear boundary between using what the blog post or tutorial describes and changing things which may break the rest of the instructions in the blog post or tutorial.

What gave me pause for thought, and what I didn't explain very well above, is that we'd then end up with two ways to link to a page with some pre-configured settings. One would be to the new project page where editing wouldn't be possible and the other would be to the main page where it would be possible. That's not necessarily a bad thing, it just made me think. The two pages would be quite similar in some ways and also quite different in others. That's really what I meant when I said that the line between them starts being a bit blurred.

**Comment by snicoll** [comment]
> What gave me pause for thought, and what I didn't explain very well above, is that we'd then end up with two ways to link to a page with some pre-configured settings.

Yes and I think we shouldn't be doing that. All the concerns I had so far regarding the use of pre-populated starters and Spring Boot generations are addressed by a dedicated page as it gives us control how to present the information to the user.

**Comment by odrotbohm** [comment]
It sounds like you were thinking quite a bit ahead of what I was trying to achieve. Which is totally fine. I'd just like to reiterate on what I would've liked to see as that seems to be less involved than some of the advanced scenarios but already enables a lot of use cases.

What I was envisioning in my original post was the ability to pre-populate form fields by using query parameters to the actual Initializr page. No direct access to the generated artifacts, no IDE integration (isn't that handled by the actual API anyway) or the like. Not even a different page as I want to be able to further customize the form.

To even step back a bit:

"As a user I'd like to be able to point to the Initializer project generation form and define certain already prepared choices, in particular a set of already selected starters and dependencies."

@michael-simons – Please weigh in here if you envisioned more than that.

In other words, I imagined the following to be possible (parameter names symbolic):

Browse to `https://start.spring.io?starters=web,jpa` and that would produce the same state as if I had browsed to `https://start.spring.io` and manually selected the web and JPA starters.

That would allow e.g. our documentation pages to point users to express something like: "If you found a bug, just create a skeleton project via this link (use expanded link here) and attach it to the bug report.". Especially when it comes to setups where multiple dependencies have to be selected to actually work (say JPA + one of the in memory databases) that would save users from finding out which minimal set of starters they need.

**Comment by michael-simons** [comment]
Wow, that's quite an effort, @philwebb in your first comment.

So first, I agree with @odrotbohm here. To rephrase it: I would like to have the main site have a linkable state.
If a certain combination of versions, languages, dependencies would not longer be possible, a 404 seems reasonable fine (with an explanation why), however a 410 (Gone) would be better. Having sensible status codes would also work for IDE developers.

Regarding the roles

**FooTech Advocate**

For me in the role as FooTech advocate, I could add such a link to the manual of our starter.
https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/blob/master/docs/manual.adoc#create-a-new-spring-boot-project.

Under "Create a new Spring Boot project" I describe what a user would have to do. To make things clear, I describe what they should and should not use in written, but also describe options for manual interaction and refer to the IDE.

I don't intend to force preconfigured projects onto people, but guide them only to chose the right dependencies.

**Experience spring developer**

Changing to the experience Spring developer, they would know about IDE integration and hopefully would read my manual.
However, they could also click on the link and open the standard site and check the nice overview of dependencies or would use the preview feature.

**The FooTech developer**

Things are hard to address here. With just restoring state, one cannot give additional information.
But that would be the same as someone telling them "please go start.spring.io".

## Having a dedicated page for permalinks ?

That being said: I assume that there already mechanisms in place to validate requests coming from IDEs or CURL, so that people cannot use invalid combinations.
They should be simple to use from the JavaScript client.

The existing page has all the UI for making changes to state. 

If that state would be encoded into the URL or with a dedicated UI for copying it, no edit buttons whatsoever would need to be added.

I would advocate for a dedicate page if my original request would have been sharable projects, which is not the case.
A sharable project is a rabbit whole, the next question I would have then would be "please make it so to include that Neo4j example here."

So, the original intention is just having a sharable state of start.spring.io, not a sharable project.

**Comment by philwebb** [comment]
Thanks for the comments. It seems like we have at least one more use-case in:

> If you found a bug, just create a skeleton project via this link (use expanded link here) and attach it to the bug report.

I think we should keep trying to drive this issue by thinking about the use-cases as much as possible. 

There are clearly a few things that the permalink page doesn't solve well:

* If the user is expected to edit the features then the extra click is annoying
* It's strange that the IDE links and curl commands are only on this page and not available from the main site.

That does seem to suggest to me that the main page and the permalink page should converge. We need to go back to the problems and think again at how we can solve them. A few thoughts:

* I don't like the URL changing has users select items. I think we should still make the link creation an explicit thing that the user does.
* I don't think we should encourage links to a specific version. I think versions should be of the form `2.1.x`.
* I think we could still have the "upgrade project" option if someone links to the project with an older version. I think this is better than linking to latest and getting a `401`. In my opinion, once a link has been created it should continue to work as long as the Boot version is still supported.
* I think the IDE information should be available on the main page. It's useful even without a link.
* I think the curl commands could also be on the main page.

**Comment by philwebb** [comment]
Proposal #2

A user wants to share a link for a pre-prepared project, they go to start.spring.io and add the dependencies that they need:

![pagep2a](https://user-images.githubusercontent.com/519772/61124123-01976d80-a49e-11e9-8413-c60276fc65c7.png)


On this page we add information about the IDE integrations, the CURL commands as well as the Generate and Browse project buttons.

Instead of clicking "Generate Project" they click "..." and pick "Share Project". This opens a small popup window a bit like the GitHub "Clone or Download" button with a URL they can copy.

The URL would be something like: `http://start.spring.io?version=2.1.x&language=java&build=maven&group=example&artifact=foo&dependencies=web,devtools,footech`

We might also offer checkboxes to change the exact query parameters we include. For example you might want to skip the group/artifact parameters.


Some important aspects are encoded here:
- The link is almost readable
- The link is to `/`
- Despite the user selecting `2.1.6` the link is for `2.1.x`
- Query parameters are used for all elements

The user can now include the link directly in their blog/documentation/bug tracker to be consumed. Clicking the link gives something like this:

![pagep2b](https://user-images.githubusercontent.com/519772/61124152-1542d400-a49e-11e9-96e4-f93510201639.png)

This page is the main start.spring.io page but with some additional information that lets the user know the form is pre-populated.

We give them:
* A quick way to start a new project
* A quick way to jump past the form and get to the "Generate", IDE, Curl etc options

**Comment by michael-simons** [comment]
I think that's a big win for many roles.

**Comment by odrotbohm** [comment]
I wouldn't even necessarily need an extra button to copy the link. If the editing of the form keeps the URL in the location bar up to date, one could already copy it from there.

**Comment by michael-simons** [comment]
I would not bike shed on the UX, but yes:

> If that state would be encoded into the URL or with a dedicated UI for copying it, no edit buttons whatsoever would need to be added.

`CMD+L` is one of those shortcuts I'm good in remembering (Selects the URL in any of the 3 big browsers).

**Comment by philwebb** [comment]
I didn't really expand my comment above where I said "I don't like the URL changing has users select items". I find the URL changing live to be quite annoying (but I don't really know why). I also want to make sure that the link we create is to the `.x` version and not an exact range.

I guess the UX experience is something that we shouldn't even attempt to design here at this point. Let's just keep focused on the problems want to solve. It doesn't really matter too much at this point how that link is created.

**Comment by odrotbohm** [comment]
What probably plays into this: I expect the users to *create* links to pre-populate the form to be way more knowledgeable about the way Initializr works than the ones solely using it. I.e. the additional button adds a bit of extra cognitive load ("What are the things I can do?") which one might argue is unnecessary or even harmful (meaning "confusing") to newbie users given that the use case of link creation is a quite an advanced one.

If you think about it: copying links from the URL bar is one of the fundamental things the web – and properly written webapps – enables. You can grab the value of the location bar and freely share it via messengers, email etc. We usually don't see "Copy the current location" buttons on other websites.

Of course, your call, but it felt like the button wouldn't necessarily need to make it into an MLP.

**Comment by odrotbohm** [comment]
> I didn't really expand my comment above where I said "I don't like the URL changing has users select items". I find the URL changing live to be quite annoying (but I don't really know why).

I think it's more annoying to bring a web app into a certain state, copying the link and – when eventually resolving it – ending up at a completely different place / state in the app. Imagine you have a look at a product on Amazon, select size and color inside the page and then trying to send that to your spouse to eventually order it.

**Comment by philwebb** [comment]
> and properly written webapps

I try not to let comments like this get to me, but I find them entirely unhelpful. I really think we should keep this about the use-cases and not what you consider to be a properly written webapp.

I personally can't think of many examples where if I type something in an edit box, the URL starts changing as I type. I think it's perfectly reasonable for the creation of a link to be an action that the user invokes.

Either way, we're moving into an area of design that I think we should steer clear of.

**Comment by odrotbohm** [comment]
Apologies, I guess I didn't completely think through what "properly" might imply. It was neither supposed to judge the authors nor was I arguing from a personal preference. It was meant as "fully aligning with the underlying concepts of the web". In this particular case meaning that *a resource, that would be useful to be addressable, actually being addressable*. This fundamental idea is what the original request boils down to.

> I personally can't think of many examples where if I type something in an edit box, the URL starts changing as I type. I think it's perfectly reasonable for the creation of a link to be an action that the user invokes.

I guess that's were we think of the problem at hand from different views. IMO the "creation of a link" is not – or: doesn't have to be – an action in the first place as users don't create links. A web application consists of a set of addressable resources so that users can bookmark them for later reference.

That ability in general is not a feature specific to a particular app. It's a fundamental trait of any web application, unless one decides to write the app in a way that breaks those, which might be a perfectly valid choice given the overall requirements – again no judgement here –, you just have to live with the fact that you then sort of having to reimplement stuff like that as a "custom feature" which then probably feels weird in turn, but for a reason.

**Comment by philwebb** [comment]
Since you made the point I've been giving it some more thought, and I think the reason that I find the URL updating so unpleasant is because we don't have an action the triggers things.

Taking your Amazon example. I go to amazon.com and search for "office chair" and hit return. See a nice char and click on it. See that they have multiple colors and click the one I want. The URL updates naturally as I click the particular action. One thing that doesn't happen is the URL updating _as_ I type "office chair".

So that's really the thing that feels most unnatural to me is the form data updating the URL as I edit it. For some things, like clicking on "Gradle" or "Maven" it feels fine. For other things, like the "Artifact ID" it's just really weird.

There might be another way to fix that. The initial form could offer a "Generate" button that doesn't give you a zip but instead redirects to a page with the download, curl, IDE commands. That page would be addressable and copy/paste from the address bar would feel natural.

I still don't like that much because we're making life hard for the 90% use-case just to make links easier. It also doesn't address the issue of `.x` in the version number. We don't want people to create a link back to a `2.1.6`, we want a link to `2.1.x` so that's different resource from the one they are currently on.

**Comment by odrotbohm** [comment]
I totally relate to that feeling. The point I was trying to make is that this weirdness is a function of the architectural choice to make all those selections on a single page. If that was a multi-resource traversing process – like your Amazon example is – the next stage would get the selections made in the first stage as input parameters and carry those forward and naturally end up in the URI.

In STS, the process of creating a project *is* a multi-step one which naturally adds those "checkpoints". I totally see why we want to avoid more than one step on the website. I think that, if exposing a URL that captures the state of some form elements is the goal, having to add a few extra bells to basically "fix" that experience is just the price we pay for that extra convenience of a single, unified form.

> I still don't like that much because we're making life hard for the 90% use-case just to make links easier.

How so? If you're not interested in capturing the state of the form, all you need to know is start.spring.io.

> We don't want people to create a link back to a 2.1.6, we want a link to 2.1.x so that's different resource from the one they are currently on.

What if it was okay to point to a particular version (i.e. simply reflect what's selected in the URI) but on resolution we drop outdated versions and use the next best one instead (instead of a `2.0.6` return `2.1.x`). I assume that some sanity checks for the versions are or need to be in place anyway. Changes made to the state of the form could be advertised in that top box you imagined in the wireframes above so that users know why the get a slightly different form than expected.

**Comment by snicoll** [comment]
Thanks all for the feedback thus far. This issue already has 22 comments and I would like us to focus the conversation on the use cases and not on particular implementation details as it is premature at this point.

Feel free to share additional thoughts on the use cases we've built so far. I'd like also to get some preliminary feedback from our UI designer and we'll report here. Thanks everyone!

**Comment by philwebb** [comment]
Good point, we're heading off into UX design again. I've hidden all the comments that don't specifically cover use-case details (including the proposals).

**Comment by snicoll** [comment]
The team has been brainstorming for quite a while and worked on different prototypes based on the feedback expressed here. We've reached a first version that we'd like users to try and we'll refine based on feedback. Thanks everyone for the input!

Closing in favour of PR #234
